src/coreclr/gc/env/gcenv.interlocked.inl

```diff
@@ -250,9 +257,11 @@ __forceinline T Interlocked::CompareExchangePointer(T volatile *destination, T e
     return (T)(TADDR)_InterlockedCompareExchange((long volatile *)(void* volatile *)destination, (long)(void*)exchange, (long)(void*)comparand);
 #endif
 #else
-    T result = (T)(TADDR)__sync_val_compare_and_swap((void* volatile *)destination, (void*)comparand, (void*)exchange);
+    static_assert(sizeof(T) == sizeof(void*), "T must be a pointer type");
```

Comment:
Use `#include <type_traits>` and 
```suggestion
    static_assert(std::is_pointer<T>::value, "T must be a pointer type");
```

Reply:
Sounds like a great idea for the next PR! 😉 (this file has many similar asserts, so I added there mostly for copy-paste-adapt to make these construct look uniform)